src/MUSI6106Exec/MUSI6106Exec.cpp

```
@@ -1,131 +1,311 @@
+        //     throw std::exception("wrong parameter of ioType (should be read / write)");
+        }
+
+        if (!phAudioFile->isOpen()) {
+            // throw std::exception("failed to open file");
+        }
+
+        if (this->fileSpec == nullptr) {
+            // throw std::exception("failed to load file spec");
+        }
+
+        phAudioFile->getLength(fileNumSample);
+
+        // shortcut for loading the full audio file
+        bool readFullFileFlag = this->blockSize < 0;
+        if (readFullFileFlag) {
+            this->blockSize = fileNumSample;
+        }
+
+        int numChannels = this->fileSpec->iNumChannels;
+        buffer = new float* [numChannels];
+        for (int i = 0; i < numChannels; ++i) {
+            buffer[i] = new float[this->blockSize];
+            memset(buffer[i], 0, sizeof(float) * this->blockSize);
+        }
+    }
 
-    clock_t                     time = 0;
+    ~AudioFileWrapper() {
+        int numChannels = fileSpec->iNumChannels;
+        for (int i = 0; i < numChannels; ++i) {
+            delete[] buffer[i];
+            buffer[i] = nullptr;
+        }
+        delete[] buffer;
+        buffer = nullptr;
+
+        if (ioType == CAudioFileIf::kFileRead) {
+            delete fileSpec;
+            fileSpec = nullptr;
+        }
+
+        phAudioFile->closeFile();
+        CAudioFileIf::destroy(phAudioFile);
+    }
 
-    float** ppfInputAudio = 0;
-    float** ppfOutputAudio = 0;
+    CAudioFileIf::FileSpec_t* getFileSpec() {
+        return fileSpec;
+    }
 
-    CAudioFileIf* phAudioFile = 0;
-    CAudioFileIf* phAudioOutputFile = 0;
+    void readData(long long numSample) {
+        if (ioType == CAudioFileIf::kFileRead) {
+            phAudioFile->readData(buffer, numSample);
+        }
+    }
 
-    CAudioFileIf::FileSpec_t    stFileSpec;
+    void writeData(long long numSample) {
+        if (ioType == CAudioFileIf::kFileWrite) {
+            phAudioFile->writeData(buffer, numSample);
+        }
```

Perhaps these should throw an exception if in the wrong mode, rather than fail silently?

---

Yeah, I was not thinking about this. I thought this class is not that important, but turn out works pretty well in making main function clean. I will make it more robust, thanks for the great point!